techmatters/hrm
# Migrate CSAM reporting data access logic to pg-promise

Primary reviewer: @stephenhand 

## Description
This PR:
- Moves the current CSAM endpoints data access logic to be implemented using pg-promise / Typescript rather than Sequelize / JS.
- Moves the code into the new HRM module layout (grouped by function).
- When a contact is created, the step of connecting the csam reports to the newly created contact is a separate step in the data access layer, instead than all being the same query. I can rollback this change but IMO being more explicit about a different table being modified in this operation makes worth the trade-off of losing efficiency in the query.

I've also done an exhaustive read on all the places where csam reports are being queried alongside the contacts, and they are queried by the FK `contactId-accountSid` in every occurrence (I don't know what I saw when raised that concern o.O Sorry!) 

After this PR, only post survey components are referencing the Sequlize based `models` :D
![Screenshot from 2022-11-17 16-43-36](https://user-images.githubusercontent.com/15805319/202543999-1177932b-1a00-41d5-b483-e6d433cae1c9.png)


### Checklist
- [x] [Corresponding issue has been opened](https://bugs.benetech.org/browse/CHI-1485)

### Verification steps
- Start local server from `master` branch, open a Flex instance, save some csams and a new contact so they are linked.
- Checkout this branch and restart the server.
- Confirm that retrieving the contact saved with the previous version still works as intended (the csam reports are bundled with the contact).
- Save another contact with csam reports and confirm that they are properly linked.
- Regression test to make sure I'm not breaking stuff.
- Make a POST to `/csamReports` endpoint, providing a valid `contactId` in the payload, and confirm that this newly created report is linked to the corresponding contact.
